File: Java/P16PowerDigitSum.java
/*
 * Project Euler Problem 16: Power digit sum
 */

import java.math.BigInteger;

public class P16PowerDigitSum {

	public static void main(String[] args) {
		int a = 2;
		BigInteger bi2 = new BigInteger(String.valueOf(a));
		BigInteger number = bi2.pow(1000);
		int sum = digitSum(number);
		System.out.println("The Power digit sum is: " + sum);
		
	}
		
	static int digitSum(BigInteger num) {
		BigInteger bi0 = new BigInteger("0");
		BigInteger sum = new BigInteger("0");
		BigInteger bi10 = new BigInteger("10");
		
		while (! num.equals(bi0)){
			sum = sum.add(num.mod(bi10));
			num = num.divide(bi10);
		}
		return sum.intValue();
	}
			
}

File: Python/p16_power_digit_sum.py
# Project Euler Problem 16: Power Digit Sum
import time

number = 2**1000

# first solution 
start = time.clock()
str(number)

sum_digits = 0

for d in str(number):
	sum_digits += int(d)

end = time.clock()

print('Time:', end-start, 'seconds')
print('Power Digit Sum:', sum_digits)

# second solution, which is much faster
start = time.clock()

sum_digits = sum(map(int, str(number)))
end = time.clock()

print('Time:', end-start, 'seconds')
print('Power Digit Sum:', sum_digits)



